content/article/enhancing-midi-hardware-with-perl.md

```diff
@@ -0,0 +1,510 @@
+— all three notes must be stopped after being played:
+
+```perl
+use constant PEDAL => 55; # G below middle C
+
+sub pedal_notes( $note ) {
+    ( PEDAL, $note, $note + 7 );
+}
+
+sub pedal_tone( $mf, $event ) {
+    my ( $ev, $channel, $note, $vel ) = $event->@*;
+    $mf->send( [ $ev, $channel, $_, $vel ] ) for pedal_notes( $note );
+    true;
+}
+
+my $mf = MidiFilter->new;
+
+$mf->add_filter( note_on  => \&pedal_tone );
+$mf->add_filter( note_off => \&pedal_tone );
+
+$mf->go;
+```
+
+We start by setting a constant containing a MIDI note value for the pedal tone.
+The sub `pedal_notes` returns this pedal tone, the played note, and its fifth.
+The callback function `pedal_tone` sends a MIDI message to output for each of
+the notes returned by `pedal_notes`.
+Note the callback yields `true` in order to prevent falling through to
+the default action.
+The callback function is applied to both the "note on" and "note off" events.
+We finish by calling the `go` method of our `MidiFilter` instance in order to
+await and process incoming messages from the keyboard.
+
+The last step is to run the script:
+
+```
+$ ./midi-filter.pl ^oxy
+```
+
+Rather than specify a fully qualified device name, we can pass in a regex which
+should match any device whose name starts with "oxy" - there is only one match
+on my system.
+
+This filter is functionally equivalent to the FL Studio Patcher patch from
+earlier, with the added benefit of being DAW-agnostic. If recording a sequence
+from this setup, all notes will be shown in the piano roll.
+
+## Feature 2 : Controller Banks
+
+The Oxygen Pro has four "banks" or sets of controls. Each bank can have
+different assignments or behaviour for the knobs, keys, sliders, and
+pads.
+
+A problem with this feature is that there is limited feedback when switching
+banks - it's not always visible on screen, depending on the last feature used.
+Switching banks does not effect the keyboard. Also, perhaps 4 banks isn't
+enough.
+
+A simpler version of this feature might be to use pads to select the bank,
+and the bank just sets the MIDI channel for all future events. There are 16
+pads on the device, for each of 16 channels. It should be more obvious which
+bank (or channel) was the last selected, and if not, just select it again.
+
+This can also be applied to the keyboard by defining callbacks for "note on"
+and "note off" (or rather, modifying the existing ones). For this device, we also
+need callbacks for "pitch wheel change" and "channel aftertouch". The callback
+for "control change" should handle the mod wheel without additional special
+treatment.
+
+The pads on this device are set up to send notes on channel 10, usually
+reserved for drums. Watching for specific notes incoming on channel 10, and
+stashing the corresponding channel should be enough to allow other callbacks to
+route events appropriately:
+
+```perl
+my $note_channel_map = {
```

Comment:
Question: How do you assign 16 channels to 7 notes here?

Maybe you want this?
```
my $initial_note = 40;
my $note_channel_map = { $initial_note++ => $_ } 0 .. 15;
```

Reply:
Yeah, it's a weird one - this is the default note layout for the pads, which is fine for finger drumming but not great as a channel switcher 🙂 

As this is to serve as an example, rather than an ergonomically sound full project, how about something like ...

```perl
sub set_channel( $mf, $event ) {
    my ( $ev, $channel, $note, $vel ) = $event->@*;
    return false unless $channel == 9;

    my $new_channel = $note - 36;
    $mf->stash( channel => $new_channel );
    true;
}
```

...with some explanation that pad notes are numbered 36..51 and this is trivially converted to the desired channel number?

Reply:
Hmm. Maybe, instead of a new algorithm, just say, "# this is the default note layout for the pads, which is fine for finger drumming"?

Reply:
It's just one additional line of code, really - we lose the hash completely so I think it's clearer.

Reply:
Sorry. I didn't realize that. Carry on! :D